src/main/scala/io/iohk/ethereum/network/TimeSlotStats.scala

```diff
@@ -0,0 +1,132 @@
+package io.iohk.ethereum.network
+
+import cats._
+import cats.implicits._
+import scala.concurrent.duration.{Duration, FiniteDuration}
+
+/** Track statistics over time a fixed size timewindow. */
+class TimeSlotStats[K, V: Monoid] private (
+    // Time resolution.
+    val slotDuration: FiniteDuration,
+    // The last written position in the buffer.
+    val lastIdx: Int,
+    // Ring buffer of slots statistics.
+    val buffer: IndexedSeq[TimeSlotStats.Entry[K, V]]
+) {
+  import TimeSlotStats._
+
+  /** Overall length of the timewindow. */
+  def duration: FiniteDuration = slotDuration * slotCount
+  def slotCount: Int = buffer.size
+
+  /** Merge new stats for a given key in the current timestamp. */
+  def add(key: K, stat: V, timestamp: Timestamp = System.currentTimeMillis): TimeSlotStats[K, V] = {
```

Comment:
as i understand `timestamp` is mainly for testing purposes ? (we do not use it in `PeerStatisticsActor`) Maybe 
`TimeSlotStats` should have dependency on `java.time.Clock` which so it could be mocked in tests, and not expose timing detail to the user.

Reply:
Yes it's a way to control time in the tests, although it could be used as a way to pass in exact observation times. I started out with `cats.effect.Clock[Id]` but it needs `Sync[Id]` which doesn't exist and then resorted to this, it's easier to pass in tests than winding a clock forward and backward. Do you have a strong preference?

Reply:
Done, using `java.time.Clock`.